test/core/end2end/tests/retry_streaming.cc

```diff
@@ -59,345 +45,129 @@ static std::unique_ptr<CoreTestFixture> begin_test(
-  cqv.Verify();
-
+  CoreEnd2endTest::IncomingMessage client_message;
+  s.NewBatch(102).RecvMessage(client_message);
+  Expect(102, true);
+  Step();
   // Client sends a second message.
-  memset(ops, 0, sizeof(ops));
-  op = ops;
-  op->op = GRPC_OP_SEND_MESSAGE;
-  op->data.send_message.send_message = request2_payload;
-  op++;
-  error = grpc_call_start_batch(c, ops, static_cast<size_t>(op - ops),
-                                grpc_core::CqVerifier::tag(3), nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv.Expect(grpc_core::CqVerifier::tag(3), true);
-  cqv.Verify();
-
+  c.NewBatch(3).SendMessage("bar");
+  Expect(3, true);
+  Step();
   // Server receives the second message.
-  memset(ops, 0, sizeof(ops));
-  op = ops;
-  op->op = GRPC_OP_RECV_MESSAGE;
-  op->data.recv_message.recv_message = &request2_payload_recv;
-  op++;
-  error = grpc_call_start_batch(s, ops, static_cast<size_t>(op - ops),
-                                grpc_core::CqVerifier::tag(103), nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv.Expect(grpc_core::CqVerifier::tag(103), true);
-  cqv.Verify();
-
+  CoreEnd2endTest::IncomingMessage client_message2;
+  s.NewBatch(103).RecvMessage(client_message2);
+  Expect(103, true);
+  Step();
   // Server sends both initial and trailing metadata.
-  memset(ops, 0, sizeof(ops));
-  op = ops;
-  op->op = GRPC_OP_RECV_CLOSE_ON_SERVER;
-  op->data.recv_close_on_server.cancelled = &was_cancelled;
-  op++;
-  op->op = GRPC_OP_SEND_INITIAL_METADATA;
-  op->data.send_initial_metadata.count = 0;
-  op++;
-  op->op = GRPC_OP_SEND_STATUS_FROM_SERVER;
-  op->data.send_status_from_server.trailing_metadata_count = 0;
-  op->data.send_status_from_server.status = GRPC_STATUS_ABORTED;
-  op->data.send_status_from_server.status_details = &status_details;
-  op++;
-  error = grpc_call_start_batch(s, ops, static_cast<size_t>(op - ops),
-                                grpc_core::CqVerifier::tag(104), nullptr);
-  GPR_ASSERT(GRPC_CALL_OK == error);
-  cqv.Expect(grpc_core::CqVerifier::tag(104), true);
-  cqv.Verify();
-
+  CoreEnd2endTest::IncomingCloseOnServer client_close;
+  s.NewBatch(104)
+      .RecvCloseOnServer(client_close)
+      .SendInitialMetadata({})
+      .SendStatusFromServer(GRPC_STATUS_ABORTED, "xyz", {});
+  Expect(104, true);
+  Step();
   // Clean up from first attempt.
-  grpc_call_unref(s);
-  grpc_metadata_array_destroy(&request_metadata_recv);
-  grpc_metadata_array_init(&request_metadata_recv);
-  grpc_call_details_destroy(&call_details);
-  grpc_call_details_init(&call_details);
-  GPR_ASSERT(byte_buffer_eq_slice(request_payload_recv, request_payload_slice));
-  grpc_byte_buffer_destroy(request_payload_recv);
-  request_payload_recv = nullptr;
-  GPR_ASSERT(
-      byte_buffer_eq_slice(request2_payload_recv, request2_payload_slice));
-  grpc_byte_buffer_destroy(request2_payload_recv);
-  request2_payload_recv = nullptr;
-
+  EXPECT_EQ(client_message.payload(), "foo");
```

Comment:
These two expectations can be moved up to be right under each of the server recv_message op completions.

Reply:
let's do that as a follow-up